base/singleton.hpp

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "base/macros.hpp"
+
+namespace base
+{
+// Simple singleton class.
+template <typename T>
+class Singleton final
```

Comment:
No, it does not, since FooImpl is not a singleton even after `using Foo = Singleton<Foo>;`